### PR TITLE
Fix: Managed access IT references outdated catalog name (#3772)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
https://github.com/DataBiosphere/azul/issues/3772

### Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

### Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

### Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks   <sub>or this PR does not require additional tasks</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR does not touch requirements*.txt</sub>

### Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate       <sub>or this PR is authored by lead</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab
- [ ] Moved issues to `prod` or `Merged prod`               <sub>or this PR does not represent a promotion</sub>

### Operator (reindex) 

- [ ] Started reindex in target deployment                  <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures             <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment              <sub>or this PR does not require reindexing</sub>
- [ ] Filed backport PR                                     <sub>or this PR does not represent a hotfix</sub>

### Operator

- [ ] Unassigned PR

### CODE REVIEW

**Review by danielsotirhos — APPROVED** [review]
Approved.

File: src/azul/__init__.py [hunk 1]
```diff
@@ -582,6 +583,15 @@ def is_integration_test_catalog(self) -> bool:
...
                 return True
 
+        @cached_property
+        def it_catalog(self) -> Optional[CatalogName]:
+            if self.is_integration_test_catalog:
+                return None
+            else:
+                catalog = self.name + self._it_catalog_suffix
+                assert self._it_catalog_re.match(catalog), catalog
+                return catalog
```
Comment (hannes-ucsc) [inline src/azul/__init__.py:593]:
```suggestion
        @cached_property
        def it_catalog(self) -> 'Catalog':
            if self.is_integration_test_catalog:
                return self
            else:
                name = self.name + self._it_catalog_suffix
                assert self._it_catalog_re.match(name), name
                return self.catalogs[name]
```

See if a fixpoint like this and returning a Catalog instance simplifies things.

File: src/azul/__init__.py [hunk 2]
```diff
@@ -557,7 +557,8 @@ class Plugin:
 
         _catalog_re = r'([a-z][a-z0-9]*(-[a-z0-9]+)*)'
         _catalog_re = r'(?=.{1,64}$)' + _catalog_re
-        _it_catalog_re: ClassVar[re.Pattern] = re.compile(_catalog_re + r'(?<=-it)')
+        _it_catalog_suffix: ClassVar[str] = '-it'
+        _it_catalog_re: ClassVar[re.Pattern] = re.compile(_catalog_re + rf'(?<={_it_catalog_suffix})')
```
Comment (hannes-ucsc) [inline src/azul/__init__.py:561]: L and technically you'd need to re.escape the interpolated string

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/__init__.py (hunk 1 above)
Comment (noah-aviel-dove) [reply src/azul/__init__.py:593]: Resolved at PL. We can't easily return instances of `Catalog`.

File: test/integration_test.py [hunk 3]
```diff
@@ -900,10 +900,10 @@ def _source_ids_from_hits(hits: JSONs) -> Set[str]:
                 with self._service_account_credentials:
                     response = self._get_url_unchecked(file_url, redirect=False)
                     self.assertIn(response.status, (301, 302))
+            # There should always be at least one managed-access source
+            # indexed and tested on the default catalog for these deployments
             elif config.deployment_stage in ('dev', 'sandbox'):
-                managed_access_catalog = 'it2'
-                assert managed_access_catalog in config.integration_test_catalogs
-                self.assertNotEqual(catalog, managed_access_catalog)
+                self.assertNotEqual(catalog, config.it_catalog_for(config.default_catalog))
```
Comment (hannes-ucsc) [inline test/integration_test.py:906]:
```suggestion
                # There should always be at least one managed-access source
                # indexed and tested on the default catalog for these deployments
                self.assertNotEqual(catalog, config.it_catalog_for(config.default_catalog))
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Comment by amarjandu** [comment]
self note: review count label was updated see https://ucsc-gi.slack.com/archives/C705Y6G9Z/p1643673935818479?thread_ts=1643673200.309169&cid=C705Y6G9Z

**Pull request merged**